### PR TITLE
define new api interface

### PR DESCRIPTION
**Pull request by yutongp**

### CODE REVIEW

File: iotex/client.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+type SendActionCaller interface {
+	Call(ctx context.Context, opts ...grpc.CallOption) (actHash string, err error)
+}
+
+type TransferCaller interface {
+	SendActionCaller
+
+	SetGasPrice(*big.Int) TransferCaller
```
Comment (zjshen14) [inline iotex/client.go]: SendActionCaller's method

File: iotex/client.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	SetGasLimit(uint64) TransferCaller
```
Comment (zjshen14) [inline iotex/client.go]: Ditto

File: iotex/client.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+type GetReceiptCaller interface {
+	Call(ctx context.Context, opts ...grpc.CallOption) (*iotexapi.GetReceiptByActionResponse, error)
+}
+
+type AuthedClient interface {
+	ReadOnlyClient
+
+	Contract(contractAddr, abi string) (Contract, error)
+	Transfer(to string, value big.Int) TransferCaller
```
Comment (zjshen14) [inline iotex/client.go]: strong type Address?

File: iotex/contract.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+type Data struct {
+	method string
+	abi    abi.ABI
+	Raw    []byte
+}
+
+func (d Data) Unmarshal(v interface{}) error { return d.abi.Unpack(v, d.method, d.Raw) }
+
+type ReadContractCaller interface {
+	SetGasPrice(*big.Int) ReadContractCaller
```
Comment (zjshen14) [inline iotex/contract.go]: Shouldn't need

File: iotex/contract.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+	SetGasLimit(uint64) ReadContractCaller
```
Comment (zjshen14) [inline iotex/contract.go]: Ditto

File: iotex/contract.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+	"context"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"google.golang.org/grpc"
+)
+
+type Data struct {
+	method string
+	abi    abi.ABI
```
Comment (zjshen14) [inline iotex/contract.go]: ABI could be way bigger than raw. Is it good to have abi be copied per call?

File: iotex/client.go (hunk 1 above)
Comment (yutongp) [reply iotex/client.go]: Cannot set payload if return sendactioncaller here

**Comment by puncsky** [comment]
why are we doing this?

File: go.sum [hunk 7]
```diff
@@ -26,6 +26,8 @@ github.com/fsnotify/fsnotify v1.4.7/go.mod h1:jwhsz4b93w/PPRr/qN1Yymfu8t87LnFCMo
 github.com/ghodss/yaml v1.0.0/go.mod h1:4dBDuWmgqj2HViK6kFavaiC9ZROes6MMH2rRYeMEF04=
 github.com/go-stack/stack v1.8.0 h1:5SgMzNM5HxrEjV0ww2lTmX6E2Izsfxas4+YHWRs3Lsk=
 github.com/go-stack/stack v1.8.0/go.mod h1:v0f6uXyyMGvRgIKkXu+yp6POWl0qKG85gN/melR3HDY=
+github.com/gogo/protobuf v1.2.1 h1:/s5zKNz0uPFCZ5hddgPdo2TK2TVrUNMn0OOX8/aZMTE=
```
Comment (zjshen14) [inline go.sum]: gogo's protobuf?

File: iotex/contract.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+type readOnlyContract struct {
+	address address.Address
+	abi     *abi.ABI
+	api     iotexapi.APIServiceClient
+}
+
+func (c *readOnlyContract) Read(method string, args ...interface{}) ReadContractCaller {
+	sender, _ := address.FromString("io1emxf8zzqckhgjde6dqd97ts0y3q496gm3fdrl6")
```
Comment (zjshen14) [inline iotex/contract.go]: What's this address?

**Pull request merged**